gpytorch/kernels/rbf_kernel_gradgrad.py

```diff
@@ -0,0 +1,157 @@
+                torch.ones(n1, n2, device=x1.device, dtype=x1.dtype).repeat(*batch_shape, 1, 1),
+            )
+            chain_rule = kp.evaluate() - outer3
+            K[..., n1:(n1*(d+1)), n2:(n2*(d+1))] = chain_rule * K_11.repeat([*([1] * n_batch_dims), d, d])
+
+            # 5) 1-3 block
+            douter1dx2 = KroneckerProductLazyTensor(
+                torch.ones(1, d, device=x1.device, dtype=x1.dtype).repeat(*batch_shape, 1, 1) / self.lengthscale.pow(2),
+                torch.ones(n1, n2, device=x1.device, dtype=x1.dtype).repeat(*batch_shape, 1, 1),
+                ).evaluate()
+
+            K_13 = (-douter1dx2 + outer1*outer1)*K_11.repeat([*([1] * (n_batch_dims + 1)), d]) #verified for n1=n2=1 case
+            K[..., :n1, (n2*(d+1)):] = K_13
+
+            K_31 = ( -douter1dx2.transpose(-1,-2) + outer2*outer2)*K_11.repeat([*([1] * n_batch_dims), d, 1]) #verified for n1=n2=1 case
+            K[..., (n1*(d+1)):, :n2] = K_31
+
+            # rest of the blocks are all of size (n1*d,n2*d)
+            outer1 = outer1.repeat([*([1] * n_batch_dims), d, 1])
+            outer2 = outer2.repeat([*([1] * (n_batch_dims + 1)), d])
+            #II = (torch.eye(d,d,device=x1.device,dtype=x1.dtype)/lengthscale.pow(2)).repeat(*batch_shape,n1,n2)
+            kp2 = KroneckerProductLazyTensor(
+                torch.ones(d, d, device=x1.device, dtype=x1.dtype).repeat(*batch_shape, 1, 1) / self.lengthscale.pow(2),
+                torch.ones(n1, n2, device=x1.device, dtype=x1.dtype).repeat(*batch_shape, 1, 1),
+            ).evaluate()
+
+            #II may not be the correct thing to use. It might be more appropriate to use kp instead??
+            II = kp.evaluate()
+            K_11dd = K_11.repeat([*([1] * (n_batch_dims)), d, d])
+
+            K_23 = ( (-kp2 + outer1*outer1)*(-outer2) + 2.*II*outer1)*K_11dd #verified for n1=n2=1 case
+
+            K[...,n1:(n1*(d+1)),(n2*(d+1)):] = K_23
+
+            K_32 = ( (-kp2.transpose(-1,-2) + outer2*outer2)*outer1 - 2.*II*outer2)*K_11dd #verified for n1=n2=1 case
+
+            K[...,(n1*(d+1)):,n2:(n2*(d+1))] = K_32
+
+            K_33 = ( (-kp2.transpose(-1,-2)+outer2*outer2)*(-kp2) - 2.*II*outer2*outer1 + 2.*(II)**2)*K_11dd + ( (-kp2.transpose(-1,-2) + outer2*outer2)*outer1 - 2.*II*outer2)*outer1* K_11dd #verified for n1=n2=1 case
+
+            K[...,(n1*(d+1)):,(n2*(d+1)):] = K_33
+
+            # Symmetrize for stability
+            if n1 == n2 and torch.eq(x1, x2).all():
+                K = 0.5 * (K.transpose(-1, -2) + K)
+
+            # Apply a perfect shuffle permutation to match the MutiTask ordering
+            pi1 = torch.arange(n1 * (2*d + 1)).view(2*d + 1, n1).t().reshape((n1 * (2*d + 1)))
+            pi2 = torch.arange(n2 * (2*d + 1)).view(2*d + 1, n2).t().reshape((n2 * (2*d + 1)))
+            K = K[..., pi1, :][..., :, pi2]
+
+            return K
+
+        else:
+            raise RuntimeError("diag=True not implemented yet")
```

Comment:
This needs to be implemented, otherwise you could get errors when e.g. preconditioning.

Reply:
I have implemented it